TACC/Core-CMS
# FP-1318: Revisit Section Pattern

## Overview

Clean up existing TACC Section pattern.

_Design has approved (via private review), but implementations should have dev review._

<details>
<summary>Wasn't There Gonna be a Section Plugin?</summary>

Yes, but it will not now, because simpler options exist:
* https://github.com/TACC/Core-CMS/pull/430
* https://github.com/TACC/Core-CMS/pull/431

</details>

Requires:
- https://github.com/TACC/Core-CMS-Resources/pull/112



## Tickets

* [FP-1318](https://jira.tacc.utexas.edu/browse/FP-1318)



## Changes

<details><summary>Simplify the code, document it better, and support automated pattern library.</summary>


<br/>

<details><summary>Move Callout title color styling to Callout stylesheet.</summary>
Though Section code is relevant, the change is to Callout element, so the code belongs with Callout. This is the logic I now follow for deciding where to place cross-pattern styles.
</details>
<details><summary>Document Section CSS.</summary>

This CSS "docblock" is required by the [automated pattern library][auto-pattern-lib].

</details>
<details><summary>Use less verbose selectors.</summary>

- (now) `.o-section` — Matches exact standalone class, so light Section **must _now_** be labeled with two classes `o-section o-section--style-light`
- (before) `[class*="o-section--"]` — Matches any modifier class, so light Section **was _before_** able to be labeled with one classes `o-section--style-light`

</details>
<details><summary>Simplify spacing code.</summary>

- space variables added unnecessary verbosity
- margin management was unnecessary
- padding management can use shorthand

</details>
<details><summary>Explain the "empty" selector code.</summary>

This code supports an odd feature specific to section intros. ([FP-1463] will extract the code.)

</details>
<details><summary>Do not hide vertical overflow on page.</summary>

- A 1px vertical overflow was on any page with 2 same-color sections.
- Instead of overlapping section atop border, make border transparent.
- And isolate updated code used to prevent borders on banners.

</details>
<details><summary>Style nested sections.</summary>

Section within a section:
- less vertical padding
- less width expansion

</details>
<details><summary>Mention FP-1462 and FP-1463.</summary>

These tickets will extract code from Section into new patterns.

</details>
<details><summary>Update Section's sample markup.</summary>

These changes support automated pattern library and section snippets.

</details>
<details><summary>Add manual pattern library snippet.</summary>

This allows creation of a snippet for consistently generating Section markup (for testing).

</details>
<details><summary>Move overflow-hiding code from templates to Core.</summary>

I am now confident that this code in global styles will not break any v1 sites (because only _horizontal_ overflow is hidden, now).

</details>


</details>



## Demos

| Demo | Notes |
| - | - |
| [__Core CMS Dev Pattern Library__][manual-pattern-lib] | __supports new `.o-section` code__ |
| [~~Automated Pattern Library~~][auto-pattern-lib] | test of [ugly KSS software][tacc-cms-kss-repo] |

[manual-pattern-lib]: https://dev.cep.tacc.utexas.edu/design-system/ui-patterns/o-section/
[auto-pattern-lib]: http://tacc-wbomar-css-pattern-library-kss.surge.sh/section-objects.html#kss-fullscreen-kssref-objects-section
[tacc-cms-kss-repo]: https://github.com/tacc-wbomar/Core-CMS-Pattern-Library-KSS



## Screenshots

| Manual Pattern Library | Frontera Home Page |
| - | - |
| ![FP-1318 Section Pattern UI](https://user-images.githubusercontent.com/62723358/150610379-60b253f8-57e6-45d6-95c5-fe6c03bce8bf.jpeg) | ![FP-1318 Frontera Home](https://user-images.githubusercontent.com/62723358/150377738-6d2c1e7f-65c2-43a3-8dc5-e37e3a2b059e.jpeg) |



## Testing

You may test locally or remote. Or do both if it makes you happy.

> __local__:  [instructions](https://github.com/TACC/Core-CMS/wiki/Test-CMS-Changes)
> __remote (core)__: https://dev.cep.tacc.utexas.edu/ ([build](https://jenkins01.tacc.utexas.edu/view/WMA%20CEP/job/Core_CMS_Build/345/), [deploy](https://jenkins01.tacc.utexas.edu/view/WMA%20CEP/job/Core_Portal_Deploy/1020))
> __remote (frontera)__: https://dev.fronteraweb.tacc.utexas.edu/ ([build](https://jenkins01.tacc.utexas.edu/view/WMA%20CEP/job/Core_CMS/346), [deploy](https://jenkins01.tacc.utexas.edu/view/WMA%20CEP/job/Core_Portal_Deploy/1021))

__Remote:__

1. Open Frontera home page on [prod](https://frontera-portal.tacc.utexas.edu/) and [dev](https://dev.fronteraweb.tacc.utexas.edu/).
    - Confirm black and white sections visually have the same _vertical_ spacing between them.
    - Confirm adjacent black sections have border between them.
    - (Optional) Notice that sections beneath banner are slightly narrower on dev than prod.[^1]
    - _Ignore broken or different images on dev._
3. Open https://dev.cep.tacc.utexas.edu/design-system/pattern-library/o-section/.
    - If this branch is deployed, then the page should load sets of Sections.
4. Follow "Remote & Local" steps.

[^1]: This is a fix to match the design. The left-and-right padding is (on most screens) 15px and comes from Bootstrap.

__Local:__

1. Be familiar with how to [Test CMS Changes](https://github.com/TACC/Core-CMS/wiki/Test-CMS-Changes).
2. Create snippets that uses template `snippets/manual-pattern-library/o-grid.html`.

    <details><summary>Snippets to Make</summary>

    > __NAME: `UI: O-Section: Light`__
    > HTML: `o-section--style-light`
    > TEMPLATE: `snippets/manual-pattern-library/o-section.html`

    > __NAME: `UI: O-Section: Dark`__
    > HTML: `o-section--style-dark`
    > TEMPLATE: `snippets/manual-pattern-library/o-section.html`

</details>

3. Follow "Remote & Local" steps.

__Remote & Local:__

1. Review the snippets used to add the content.
2. Confirm that each class does what [each class says it should](https://github.com/TACC/Core-CMS/blob/task/FP-1318-section-pattern-revisit/taccsite_cms/static/site_cms/css/src/_imports/objects/o-section.css#L25-L26).



## Notes

The "sub" patterns of intro, banner, and layout will be extracted in [FP-1462] and [FP-1463].

<details><summary>To Do</summary>

Design Tweaks
- [x] nested section width should extend past parent (example:  `95vw`, but find a way to still use `100vw`) — 210ffe4
- [x] nested section should have some top/bottom padding — 210ffe4
- [x] fix `[class*=o-section--style]+[class*=o-section--style]` to be color A after color A (**never** B+A nor A+B) — b682ded
- [x] note in a ticket that gray section style will have the same top/bottom spacing as light and dark — [FP-1470](https://jira.tacc.utexas.edu/browse/FP-1470)

</details>

[FP-1462]: https://jira.tacc.utexas.edu/browse/FP-1462
[FP-1463]: https://jira.tacc.utexas.edu/browse/FP-1463
